Improve version handling to avoid multiple puppet runs for some situa…

bdellegrazie opened the pull request:
…tions

When an administrator knows the version of collectd that will be used or
at least the minimum version available the need for two puppet runs before
convergence can be avoided or at least minimised.

Instead of using the fact in the templates they now use a class variable
set to one of (in priority order):
* collectd_version (i.e. the fact)
* version (the semver matched part of it only)
* minimum_version (undef by default)

Existing behaviour is preserved except for a corner case where version is
set to something specific and collectd is not yet installed. In this case
puppet will only take one run and assume the version specified when creating
the templates

references pdxcat/puppet-module-collectd#162

blkperl commented:
PIng @txaj for review

blkperl commented:
Is it possible to write a spec test to test the new behavior?

bdellegrazie commented:
Hi @blkperl, I'd like to add tests for the new functionality but I'm very new to puppet-rspec and I'm not sure how. Specifically I'm having difficulty accessing the `$_collectd_version` created in `init.pp` as its essentially being used as a module-wide variable in the templates.
Do you have any suggestions?

bdellegrazie commented:
ping @blkperl and @txaj - any update or further comments?

txaj commented on README.md:
```diff
@@ -1087,7 +1092,10 @@ See metadata.json for supported platforms
+###Puppet needs two runs to correctly write my conf, why?
+
+Some plugins will need two runs of Puppet to fully generate the configuration for collectd. See [this issue](https://github.com/pdxcat/puppet-module-collectd/issues/162).
+This can be avoided by specifying an explicit version (`$version`) or a minimum version (`$minimum_version`) to the collectd class
```
Could you be more explicit about how this parameter works ? like "setting this parameter to 9.99 will makes this module assume at the first run (when the fact responsible to provide the collectd version is not yet avaibable) that your systems are running collectd 9.99, and generate the configuration accordingly"

bdellegrazie replied:
Sure will do

txaj commented on manifests/init.pp:
```diff
@@ -14,12 +14,16 @@
   validate_array($include, $typesdb)
 
+  # Version for templates
+  $_collectd_version = pick($::collectd_version, regsubst($version, '^\d+(?:\.\d+){1.2}', '\0'), $minimum_version)
```
what is this regex for ? if that's for sanity, I'd prefer a validate_re so we can fail if the input is incorrect

txaj replied:
this changelog hurts. Instead of changing collectd_version everywhere, why not change the name of the fact ? ping @blkperl do you see any argument to say that this could be a bad idea ?

bdellegrazie replied:
The regex is only meant to match a version number of collectd (as much of the major, minor and patch parts that are supplied). it returns undef if version happens to be 'installed', 'present' or 'absent'. validate_re is not appropriate in this circumstance.

re: fact name
I don't mind changing the fact name if that's what is preferred. I changed the templates because I was concerned that the fact was considered a part of the modules 'public' interface.
Make a decision either way and let me know.

txaj commented:
@bdellegrazie for testing I only see one way : take a module that generates a different configuration depending on the collectd version, and test the output when you set the fact to undef/value1/value2 and the default to undef/value1/value2 and check that it makes sense

bdellegrazie commented:
@txaj, @blkperl: I've updated the PR to do the following:
* renamed collectd_version fact to collectd_real_version
* corrected an error with the regexp which selects the version for the templates.

I had to use a test module in the fixtures dir to be able to create the tests around the collect_version but they do work.

igalic commented:
thanks @bdellegrazie and @txaj! i've merged this manually.

igalic commented:
since merging this, we now have a [failing test](https://github.com/puppet-community/puppet-collectd/blob/master/spec/classes/collectd_init_spec.rb#L73-L77)

@bdellegrazie do i understand it correctly, that we should be checking for collectd_real_version here (as well as in the code)?

bdellegrazie commented:
yes the fact should be collectd_real_version - I thought it was on my
branch?

On 28 September 2015 at 16:15, Igor Galić <notifications@github.com> wrote:

> since merging this, we now have a failing test
> <https://github.com/puppet-community/puppet-collectd/blob/master/spec/classes/collectd_init_spec.rb#L73-L77>
>
> @bdellegrazie <https://github.com/bdellegrazie> do i understand it
> correctly, that we should be checking for collectd_real_version here (as
> well as in the code)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppet-community/puppet-collectd/pull/305#issuecomment-143773346>
> .
>



-- 
Kind regards,

Brett Delle Grazie

bdellegrazie commented:
Okay, I realise someone has added more tests (great!) but when merging the
fact needs to be collectd_real_version - then the internal
'collectd_version' variable that is used by the templates will resolve
correctly.

On 28 September 2015 at 16:41, Brett Delle Grazie <
brett.dellegrazie@gmail.com> wrote:

> yes the fact should be collectd_real_version - I thought it was on my
> branch?
>
> On 28 September 2015 at 16:15, Igor Galić <notifications@github.com>
> wrote:
>
>> since merging this, we now have a failing test
>> <https://github.com/puppet-community/puppet-collectd/blob/master/spec/classes/collectd_init_spec.rb#L73-L77>
>>
>> @bdellegrazie <https://github.com/bdellegrazie> do i understand it
>> correctly, that we should be checking for collectd_real_version here (as
>> well as in the code)?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/puppet-community/puppet-collectd/pull/305#issuecomment-143773346>
>> .
>>
>
>
>
> --
> Kind regards,
>
> Brett Delle Grazie
>



-- 
Kind regards,

Brett Delle Grazie

igalic commented:
@bdellegrazie did i do a poor job of rebasing / merging your branch?

bdellegrazie commented:
@igalic I'd hardly count two functions as a poor merge :)
No, there is however something 'broken' with Puppet 4.0 according to Travis output from PR #343.
I don't know how to fix this one however.
